Add archival logic to MSRuleCleaner.

todor-ivanov opened the pull request:
Fixes #10018 

#### Status
ready

#### Description
Moves the workflows archival logic from the CherryPy thread to the MSRuleCleaner. This is implemented in two steps:
   * Cleans all the LogDB records for the current workflow
   * Makes the needed request status transition in reqmgr   

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
Needs some more configuration parameters such as:
`wmstatsUrl`, 
`logDBUrl`,
`logDBReporter`,
`archiveDelayHours`

The related deployment config PRs:
https://github.com/dmwm/deployment/pull/1010
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/54
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/merge_requests/55



#### External dependencies / deployment changes
NO

todor-ivanov committed:
Add archival logic to MSRuleCleaner.

Add IsArchDelayExpired to MSRuleCleanerWflow

First code review iteration.

Remove archival code from CherryPyThreads.

Get status transition history from reqmgr.

Fix bug for not traversing archival tests && double check LogDB after cleaning.

Fix double assignment for logDB.

Second code review iteration && Bugfix - worngly set return value from _checkLogDBClean.

Splitting MSRuleCleanerArchival exception in two separate ones.

Final code review iteration && Moving the call to _cleanLogDB() inside setLogDBClean().

todor-ivanov commented:
Alan @amaltaro please go ahead and take a look, while I prepare a fix for the Unit test that I expect to break with this PR, so we may discuss the so proposed solution here.  Thanks in advance!

todor-ivanov commented:
test this please

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -60,9 +64,14 @@ def __init__(self, msConfig, logger=None):
         self.emailAlert = EmailAlert(self.msConfig)
         self.curlMgr = RequestHandler()
+        self.wmstatsSvc = WMStatsServer(self.msConfig["wmstatsUrl"], logger=self.logger)
+        self.logDB = self.logDB = LogDB(self.msConfig["logDBUrl"],
```
buggy :)

todor-ivanov replied:
@amaltaro excuse my ignorance here ;), but please do elaborate!

amaltaro replied:
repeated assignment

todor-ivanov replied:
Aaaand... this is a silly typo!!!  Fixing it now.
Thanks!

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
         wflow['IsClean'] = self._checkClean(wflow)
         return wflow
 
+    def _checkLogDBClean(self, wflow):
```
I would call it `isLogDBClean`. The docstring needs some working IMO.

todor-ivanov replied:
Ok, sorry again. I will need some more info about what exactly do you think needs to be worked on in the docstring. I will try to make it as clear as possible. 
About the name: I was following a pattern here and all private methods that are not supposed to be called independently are returning a `bool` variable instead of a `wflow` object and all of them I was prefixing them with the same substring `_check....` But if you think I should break the pattern I will rename it.

todor-ivanov replied:
In general I tend to create a separate method when a separate call just to the check method is expected to be made somewhere else into the code. In all other cases, I do agree with you that having the logic enclosed inside the function that is about to change the flag status itself is the best. I will see what I can do with the `_checkLogDB` method.

amaltaro replied:
Thanks!

amaltaro replied:
Marking it as resolved! Thanks

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        """
+        # Check the available status transitions before we decide the final status
+        targetStatusList = RequestStatus.REQUEST_STATE_TRANSITION.get(wflow['RequestStatus'], [])
+        targetStatusReg = re.compile(r'.*archived')
```
How about making it an object variable and compile it only once?

todor-ivanov replied:
I was also thinking the same and hesitating initially. But now I am sure. 
Done now.

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+            self.logger.info("Deleting %s from LogDB WMStats...", wflow['RequestName'])
+            res = self.logDB.delete(wflow['RequestName'], agent=False)
+            if res == 'delete-error':
+                self.logger.error("Failed to delete logDB docs for wflow: %s", wflow['RequestName'])
```
With a `delete-error` response, you still report a cleanStatus=True here.

todor-ivanov replied:
That was a good catch... how did I miss that myself?!
Let me see how I can fix it

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+
+    def _checkArchDelayExpired(self, wflow):
+        """
+        A function to check Archival Expiration Delay based on the inromation
```
There is a typo here. And even though it's just a docstring, I'd suggest not to user capitals where it's not needed, like "Archival Expiration Delay"

todor-ivanov replied:
Thanks! 
Fixed

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        currentTime = int(time.time())
+        threshold = self.msConfig['archiveDelayHours'] * 3600
+        outputMask = ['RequestTransition']
+        inputConditon = {'RequestName': wflow['RequestName']}
```
This will make one call per workflow, but we should actually make a single call per status, and simply go through the workflows.

todor-ivanov replied:
Ok, this one actually took me longer than I would like spend but anyway I think I managed to solve it in the best possible manner. I had to sacrifice memory footprint for run time complexity, but at the end I figured out that we can afford it.  @amaltaro please take a look and give me your opinion.

todor-ivanov replied:
This was remade yet again. To use `reqmgr` instead of `wmstatsServer` as @amaltaro was suggesting, and is actually a better looking piece now. Resolving this conversation now.

amaltaro commented on src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWflow.py:
```diff
@@ -70,6 +70,8 @@ def docSchema(self):
             ('PlineMarkers', None, list),
             ('IsClean', False, bool),
+            ('IsLogDBClean', False, bool),
+            ('IsArchDelayExpired', False, bool),
```
I would completely write this name, such as `IsArchivalDelayExpired`

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        :param wflow:       A MSRuleCleaner workflow representation
+        :return:            Bool
+        """
+        archiveStatus = False
```
I somehow have the feeling that we are overloading the usage of "private" methods in this module, basically every method has its own private method to execute the action. IMO, I'd move this logic under the `archive` method itself.

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        elif not wflow['IsArchDelayExpired']:
+            msg = "Archival delay period has not yet expired for: %s." % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+        elif not self.msConfig['enableRealMode']:
```
I guess you simply want to log a DRY-RUN thing here?

todor-ivanov replied:
Actually no. If we do not leave the function properly with raised exception here, then the workflow is going to get archived. 
But if you simply mean that the Error message should to be changed to "DRY-RUN  mode .... ", then I agree, that should be better.

amaltaro replied:
I see. However, we should distinguish the following two use cases:
1) workflow cannot be archived because there was an unexpected problem (e.g. wrong status, not all docs successfully cleaned, etc)
2) workflow is in an expected state (e.g. archival delay hasn't been met, service is in dry-run).

In addition to that, errors that are not really errors shouldn't be logged as error. What do you think?

todor-ivanov replied:
I totally agree. And the only way out of this I see if we put a second exception there and (re)name them both accordingly (Honestly speaking I do not like the name of the current exception either - it does not give any clue that this is actually an exception or an error of some kind)

todor-ivanov replied:
Done

amaltaro commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,139 @@ def setClean(self, wflow):
+        elif not wflow['TargetStatus']:
+            msg = "Could not determine which archival status to target for: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+        elif not wflow['IsLogDBClean']:
```
I don't understand why raising exceptions here. Same for the two block belows (IsArchDelayExpired and enableRealMode). Shouldn't we simply log (if not already) an error and move to the next workflow?

todor-ivanov replied:
As I started to explain in the previous section we must exit the function in order to avoid actual status transition to one  of `*-archival` statuses. And the best way to do this is with an exception, because that way we can signal the upper level function not to update any of the relevant counters regarding the failing worklow.

amaltaro reviewed (changes requested):
Todor, it's going in the right direction.
However, I have made some minor and major comments along the code.

In addition to that, I'd suggest the following:
* docstrings (especially the return: documentation) need to be worked further. I'm not sure we need that extra spacing as well..
* please make a 2nd commit - in this same PR - with the archival removal from the ReqMgr2 CherryPy thread. This has to be an atomic operation to avoid issues, and the way to have it is by providing everything in the same PR.

todor-ivanov commented:
@amaltaro just take a look at the currently proposed changes. Mostly speaking about the in memory approach for the status transition history. This is not the final version yet. There are still the changes to the cherrypy thread to come.

todor-ivanov commented:
test this please

todor-ivanov committed:
Fix unit tests

Fix unit tests

Fix unit tests

Fix unit tests

Fix unit tests - Add RequestTransition && Split MSRuleCleanerArchival exception.

Fix unit tests - Fix counters && Add assertRaises to cleaned workflows test too.

Fix unit tests - Fix expectedWflow

todor-ivanov commented:
test this please

amaltaro reviewed (changes requested):
Todor, this is a very incomplete review because I wanted to give you some guidance on the way you are fetching workflow information here.

The `_buildStatusHistory` method is actually not necessary, because the information retrieved there is a subset of what it already retrieves through `getRequestRecords` method.
However, I would actually change the service we fetch workflow information from here, instead of using reqmgr2, I'd suggest to use wmstatsserver which has all that cached. In addition to that, it looks like you are only interested in 4 or 5 workflow properties, so maybe we could use this `getFilteredActiveData` wmstats method and fetch only the information we need. This MS polling cycle is very large, so it won't be too much of a load on the wmstatsserver.

Ping me on slack if you'd like to discuss it further.

todor-ivanov commented:
@amaltaro Alan, Please take a look at this PR again. I managed to also catch two bugs and cover them with my latest commit. There will still be two unittests that are about to break, but this is because I need to figure out how to deal with the LogDB session inside the unit test setup. Anyway, please have a look.

ps. In case it will be easier fro you just tell me and I will sqaush all the commits in two sets - one for the feature changes and one for the unit tests (this PR kind of became heavy to track already)

amaltaro commented:
>... but this is because I need to figure out how to deal with the LogDB session inside the unit test setup ..

Todor, I think I have already struggled with LogDB in unit tests as well. Try setting it to `None` (or have a look at other unit tests)

amaltaro commented on src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWflow.py:
```diff
@@ -70,7 +70,10 @@ def docSchema(self):
             ('IsClean', False, bool),
-            ('ForceArchive', False, bool)]
+            ('IsLogDBClean', False, bool),
+            ('IsArchivalDelayExpired', False, bool),
```
Given that `IsArchivalDelayExpired` depends solely on the request dictionary, I'm inclined to say that moving that code over to this module would make things more clear.

todor-ivanov replied:
At first glance I was about to agree with you immediately. But once I actually tried to implement it, I immediately ran into the need to import the `msConfig` object into `MSRuleCleanerWflow.py`, which makes things a little bit more complicate. And this made me immediately remember that we decided to keep this one as simple/generic as possible, but not putting a lot of weight to this class similar to `MSOutputTemplate.py`.

todor-ivanov replied:
I do agree with you that keeping as little as possible in the workflow representation object would be kind of a cleaner approach, but I am not quite sure how much of an impact would that have on the overall memory footprint from a single run. Because we are basically re-initializing the wflow object in place with every new workflow (see here [1]). And the memory should be release right after we get out of scope - which I expect to happen once the `_dispathWflow` function returns a value. And given the fact that we are feeding it a wflow at a time I thought the increased memory consumption while keeping the whole transition history of a single workflow would be neglectable compared to the enormous and detailed list of workflows returned by `reqmgr` and which we keep in memory until the end of the whole cycle (this is when this one gets out of scope). Of course all that  needs to be confirmed with a full run of a proper profiler, which I need to admit I haven't done - so there is a possibility that I may be wrong here. 

So I would suggest we  continue now as it have been already tested, and we may come back to this once we start thinking (eventually) about feeding the whole service in a different way. Will keep the conversation not resolve though - just for a future reminder.   

[1]
https://github.com/dmwm/WMCore/blob/875c3c5a48d0d82df947cdf13466c469c33ca27a/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py#L188-L190

amaltaro replied:
And here is where you contradict yourself :-), compared to your previous comment:
"""
And this made me immediately remember that we decided to keep this one as simple/generic as possible, but not putting a lot of weight to this class similar to MSOutputTemplate.py.
"""

I try to push for workflow-related logic inside the workflow class, and the argument you give me is that you're concerned with the memory footprint...

The approach I would adopt is to add all the workflow logic in the same place, same as I've done for the workflow object used in MSTransferor:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/DataStructs/Workflow.py

Another good candidate for this would be a helper function that internally evaluates whether the workflow can be archived, e.g.:
```
if myWflowObj.canArchive():
    blah
```
and then that `canArchive` method would be implemented in this MSRuleWflow class and evaluate all the flags we need to evaluate.

Anyhow, I just wanted to make clear that my proposal was to inject workflow logic to the workflow struct class, not to make it as simple as possible and leave logic to be implemented elsewhere. As you said, this code works and it has been tested, let's move on!

todor-ivanov replied:
Agreed!

amaltaro reviewed (changes requested):
Sorry Todor, I forgot publishing half of the review I had made yesterday. Here it goes a few more comments and suggestions

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,124 @@ def setClean(self, wflow):
+        wflow['IsLogDBClean'] = self._checkLogDBClean(wflow)
+        return wflow
+
+    def findTargStatus(self, wflow):
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,124 @@ def setClean(self, wflow):
+        """
+        cleanStatus = False
+        try:
+            self.logger.info("Deleting %s from LogDB WMStats...", wflow['RequestName'])
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,124 @@ def setClean(self, wflow):
+            if res == 'delete-error':
+                self.logger.error("Failed to delete logDB docs for wflow: %s", wflow['RequestName'])
+                return cleanStatus
+            self.setLogDBClean(wflow)
```
So in this case we do not fetch the document from a couchdb view, but from the logDb instance itself, and that's  why I was expecting the change to be kind of instant. The additional check was done only because I did not want to have some leftovers behind, but I agree - In case we are  sure that once the deletion call is properly sent then the documents are 100% about to be deleted, then yes this additional call to the database would be redundant.

todor-ivanov replied:
BTW I am about to make the change you requested here. Lets see how it goes.

todor-ivanov replied:
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +370,124 @@ def setClean(self, wflow):
+            self.logger.debug("Could not find status transition history for %s", wflow['RequestName'])
+        return archDelayExpired
+
+    def setArchDelayExpired(self, wflow):
```
Done

todor-ivanov commented:
Hi @amaltaro, Please go ahead and review the final version of this PR. I think I have addressed all the comments from the previous iteration.  And also the unit tests must be fixed now.

amaltaro reviewed (commented):
Todor, code is looking almost ready to me. I had a few minor comments along the code and one - about cleaning LogDB - that might require some further work, but isn't a showstopper though.

In addition to that, please have a look at the PR report from jenkins. There are a few very easy improvements that can be made here (not only the bold ones).

Last but not least, it came to my mind today that we might want to decrease the polling cycle in the future (for a quicker workflow archival). I wonder if we make Rucio calls even if the archival delay hasn't been reached yet?

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -139,6 +160,7 @@ def execute(self, reqStatus):
             requestRecords = {}
             for status in reqStatus:
                 requestRecords.update(self.getRequestRecords(status))
+                # self._buildStatusHistory(status)
```
Correct. Thanks!

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -60,9 +73,14 @@ def __init__(self, msConfig, logger=None):
         self.mode = "RealMode" if self.msConfig['enableRealMode'] else "DryRunMode"
         self.emailAlert = EmailAlert(self.msConfig)
         self.curlMgr = RequestHandler()
+        self.targetStatusReg = re.compile(r'.*archived')
```
Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +384,123 @@ def setClean(self, wflow):
+        A function to set the 'IsLogDBClean' flag based on the presence of any
+        records in LogDB for the current workflow.
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
```
Agreed and Done

todor-ivanov commented on src/python/WMCore/MicroService/Unified/MSRuleCleaner.py:
```diff
@@ -358,23 +384,123 @@ def setClean(self, wflow):
+            msg = "Could not determine which archival status to target for: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchivalError(msg)
+        if not wflow['IsLogDBClean']:
+            if not self._cleanLogDB(wflow):
```
Agreed, I was feeling the same way while putting it here, but couldn't come up with a better place though. I will see what I can do about it now.

todor-ivanov replied:
Done now.

todor-ivanov commented:
@amaltaro , I just commited one possible way of reshuffling some function calls so that we have the `_cleanLogDB()` called outside the archival method. Please take a look and let me know if you like it, so I can squash 'em all ;)

amaltaro commented:
Thanks for all the follow up, Todor.
